akka-persistence/src/main/scala/akka/persistence/journal/AsyncWriteJournal.scala

```diff
@@ -93,30 +94,29 @@ trait AsyncWriteJournal extends Actor with WriteJournalBase with AsyncRecovery {
       }
 
     case r @ ReplayMessages(fromSequenceNr, toSequenceNr, max, persistenceId, persistentActor) ⇒
-      // Send replayed messages and replay result to persistentActor directly. No need
-      // to resequence replayed messages relative to written and looped messages.
-      asyncReplayMessages(persistenceId, fromSequenceNr, toSequenceNr, max) { p ⇒
-        if (!p.deleted) // old records from 2.3 may still have the deleted flag
-          adaptFromJournal(p).foreach { adaptedPersistentRepr ⇒
-            persistentActor.tell(ReplayedMessage(adaptedPersistentRepr), Actor.noSender)
-          }
-      } map {
-        case _ ⇒ ReplayMessagesSuccess
-      } recover {
+
+      asyncReadHighestSequenceNr(persistenceId, fromSequenceNr).flatMap { highSeqNr ⇒
+        val toSeqNr = math.min(toSequenceNr, highSeqNr)
+        if (highSeqNr == 0L || fromSequenceNr > toSeqNr)
```

Comment:
I went through the entire flow now. You're right that `fromSequenceNr` comes from the snapshot store. 

The case you mention will not happen however:
- recover(to = 13), results in:
  - load snapshot(to 13), so the snapshot store would not yield the 17 snapshot, it may yield a lower one or None
- a) we get `snapshotOffer = None` and `to = 13` => `snapshotOffer = None` is translated to a `from = 0`, start replay from 0 => OK
- - a) we get `snapshotOffer = Some(5)` and `to = 13` => `snapshotOffer` bumps the fromSequenceNr, `from = 5 + 1`, start replay from 6, since snapshot is applied => OK

So the case when the fromSequenceNr is lower than toSeqNr is when:
- user: recover to 18
- snapshotstore: snapshot offer available for 14
- recovery continues `fromSequenceNr = 15`
- journal: here a possibility for a mismatch, it could yield: `asyncReadHighestSequenceNr => 12`
  - why could this happen: because the journal is somehow out of sync with the snapshot store. For example someone retrored a old backup of the journal, but NOT of the snapshot store, so it has latter sequence ids
- weird state: if we say it's Future.successful then the actor is effectively in seqNr 15 state, from the snapshot offer...

This may actually be OK now that I have written it out, since the snapshot will be offered anyway always, so we'll get this state back again each time, and the sequenceNr does not overrwrite anything, there's a gap in the journal, but we don't hit it since we use the snapshot...

I think it would break if we would recover in such situation without using the snapshot. But perhaps that's expected.